scripts/lint-linkcheck.mjs

```diff
@@ -181,12 +250,55 @@ class BrokenLinkChecker {
 	pathnameToHtmlFilePath (pathname) {
 		return path.join(this.buildOutputDir, pathname, 'index.html');
 	}
+
+	/**
+	 * Attempts to find a Markdown source file for the given `pathname`.
+	 * 
+	 * Example: Given a pathname of `/en/some-page` or `/en/some-page/`,
+	 * searches for the source file in the following locations
+	 * and returns the first matching path:
+	 * - `${this.pageSourceDir}/en/some-page.md`
+	 * - `${this.pageSourceDir}/en/some-page/index.md`
+	 * 
+	 * If no existing file is found, returns `undefined`.
+	 */
+	tryFindSourceFileForPathname (pathname) {
+		const possibleSourceFilePaths = [
+			path.join(this.pageSourceDir, pathname, '.') + '.md',
+			path.join(this.pageSourceDir, pathname, 'index.md'),
+		];
+		return possibleSourceFilePaths.find(possiblePath => fs.existsSync(possiblePath));
+	}
+
+	/**
+	 * Attempts to find the given link `href` inside `input` and returns its index on a match.
+	 * 
+	 * Prevents false positive partial matches (like an href of `/en/install` matching
+	 * an input containing `/en/install/auto`) by requiring the characters surrounding a match
+	 * not to be a part of URLs in Markdown.
+	 */
+	indexOfHref (input, href) {
+		let i = input.indexOf(href);
+		while (i !== -1) {
+			// Get the characters surrounding the current match (if any)
+			let charBefore = input.substring(i - 1, i);
+			let charAfter = input.substr(i + href.length, 1);
+			// If both characters are not a part of URLs in Markdown,
+			// we have a proper (non-partial) match, so return the index
+			if ((charBefore + charAfter).match(/^[\s"'()[\],.]*$/))
```

Comment:
This wouldn't match an absolute URL right? So if someone linked to `https://docs.astro.build/404`, we'd fail to annotate it.

Reply:
The regex is there to check the character immediately before and after the broken link, not the broken link URL itself. So if your broken link is contained in a markdown file like `See the [Install Guide](https://docs.astro.build/404) for more details.`, the regex will look at the opening and closing round brackets and succeed, making it properly annotate it.

The only reason this check is in there is to prevent partial URL matches from occuring in which case it would annotate lines that don't contain an error. For example, if `https://docs.astro.build/en/install` is a broken link, and `https://docs.astro.build/en/install/auto` is not, we do not want to annotate the perfectly valid `https://docs.astro.build/en/install/auto` just because a substring matches the broken link.

Reply:
Ahhh, OK. It's just that the logging for the checker always outputs the full absolute URL, so I was assuming you then were converting that to a path. (So in my example I thought you might be looking for `/404` when a file actually includes the absolute URL.)